[JENKINS-57761] - Add abstract test for export/configure roundtrip

MRamonLeon opened the pull request:
<!--
To mark your pull request as work in progress put the construction emoji 🚧 in your title. (hint: copy it)
Helps us to block accidental merges of unfinished work.
-->

### Your checklist for this pull request

🚨 Please review the [guidelines for contributing](../blob/master/docs/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!

- [x] Please describe what you did

- [x] Link to relevant GitHub issues or pull requests
https://github.com/jenkinsci/metrics-plugin/pull/45 All code in the metrics plugin
https://github.com/jenkinsci/metrics-plugin/pull/46 The implementation in the metrics plugin

- [x] Link to relevant [Jenkins JIRA issues](https://issues.jenkins-ci.org)
[JENKINS-57761](https://issues.jenkins-ci.org/browse/JENKINS-57761) Developed while creating the roundtrip test for metrics plugin.

- [x] Did you provide a test-case? That demonstrates feature works or fixes the issue.

<!--
Put an `x` into the [ ] to show you have filled the information below
Describe your pull request below
-->

### Description

Add a new abstract test to be used by plugins as a full export / configure test. The goal is to make as easy as possible the life to plugins authors and at the same time, have tested as many plugins as possible. With this PR, plugin authors only need to extend a class and override 3 methods. See https://github.com/jenkinsci/metrics-plugin/pull/46

Currently, there is a problem exporting the full configuration and importing it. The _**defaultProperties**_ are exported and make the file exported to be not importable. See https://issues.jenkins-ci.org/browse/JENKINS-57122 I've done a workaround by using the config piece of the plugin under test, but it should be removed once the problem is solved.

timja commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,254 @@
+    public LoggerRule logging = new LoggerRule();
+
+    static {
+        // The restartable rule add spaces to Jenkins home and it makes the second and next steps to fail loading the
```
is this something that should be fixed?

MRamonLeon replied:
Fixed

timja commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,258 @@
+
+            // Get the full configuration
+            //String jenkinsConf = getJenkinsConfViaWebUI();
+            //hack: the full Jenkins config fails due to defaultProperties of maven, we use the config of the plugin
```
there's a PR for it but seems unlikely to be merged ☹️
may need to hack something into jcasc to blacklist it
https://github.com/jenkinsci/jenkins/pull/3992

MRamonLeon replied:
I've added a workaround, but would it be great if it works. We then can do the full export / replace roundtrip :smiley:

timja replied:
Without that working I'm not that comfortable with this PR, i.e if a plugin user implements this test then there's no coverage for the exporting of their plugin.

The the bit which seems to be missing is without asserting the export is what you expect you risk the plugin author:
1. not viewing the export and realising it's not what they expect
2. unintentionally breaking the config in an update but not realising because the round trip works

MRamonLeon replied:
Hi Tim. I've tried to check the exported configuration has the same values as the YAML used. But, again, I faced a bug and I've commented out the code.

When running Jenkins normally:
You can add secrets ciphered in the YAML. CasC uploads it without reciphering. And when you export the configuration, the data is the same as it was in the yaml file.

When running with mvn running tests or hpi:run:
CasC ciphers the secrets, no matter if they were already ciphered in the yaml.

This bug makes impossible for plugins with secrets in the yaml to pass the test. We can solve this issue to uncomment this hack, or fix the full export to uncomment both hacks I did. :-)

Anyway, I've pushed the code.

MRamonLeon replied:
I've added the RoundTripMailerTest to prove it's working fine (when no secrets in the yaml)

casz commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,258 @@
+/**
+ *
+ */
+public abstract class ExportImportRoundTripAbstractTest {
```
Could you explain the reasoning for a restartable Jenkins with roundtrip configuration? 
What's the issue? What are you trying to solve?
Why can't you use the existing test method that others have done?

MRamonLeon replied:
I'm trying to test easily for a plugin author:
1. The instance is configured with his/her piece of yaml
2. The full configuration exported can be applied (not currently - added a hack)
3. The instance is configured using the default config file.

fcojfernandez replied:
I think the idea is great. I would change maybe some points of the implementation since I've said in the general comment, but that should help to any maintainer to test the compatibility of their plugins.

MRamonLeon commented on integrations/src/test/java/io/jenkins/plugins/casc/GlobalMatrixAuthorizationTest.java:
```diff
@@ -1,35 +0,0 @@
- * @author Mads Nielsen
- * @since TODO
- */
-public class GlobalMatrixAuthorizationTest {
```
I deleted these two files by mistake, sorry for the confusion

MRamonLeon commented:
@oleg-nenashev @batmat @alecharp @fcojfernandez @varyvol A useful abstract test to go faster when checking plugins compatibility with JCasC.

varyvol commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,260 @@
+                // Verify the configuration set at home/jenkins.yaml is loaded
+                assertConfiguredAsExpected(r, resourceContent);
+            } finally {
+                restoreJenkinsConfig();
```
What if the failure happens during the first step?

fcojfernandez replied:
My understanding is that each execution should have its own Jenkins Rule running, so it should not impact. The test only would fail. Am I right @MRamonLeon ?

MRamonLeon replied:
You're right. But the `System.setProperty` may affect other tests. Anyway, I've removed the setter of the property after other fixes because it was failing from the beginning and it seems to work fine. So, after removing that I can safely remove the backup and restore code.

fcojfernandez commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,260 @@
+     * @throws IOException If an exception is thrown managing resources or files.
+     */
+    @Test
+    public void exportImportRoundTrip() throws IOException {
```
nit: I've seen similar tests named as `configRoundtrip` in the past. Doesn't it suit here?

fcojfernandez commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,260 @@
+     */
+    public abstract String stringInLogExpected();
+
+    @Before
```
I think it can be dangerous. If someone extends this class and adds other tests that are not related with the subject of this abstract class, this method would be executed always. Maybe it's worth to call it from the test method

fcojfernandez commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,260 @@
+    Annotated with @After is executed before the second step is ended.
+    We need to call it at the end of the second step.
+     */
+    //@After
```
If it's not used, better remove it?

timja commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,224 @@
+     * class, then: my-config.yaml
+     * @return the resource name and path.
+     */
+    protected abstract String configResource();
```
can't this be defaulted?
```suggestion
    protected String configResource() {
        return "configuration-as-code.yaml";
    }
```

MRamonLeon replied:
Could be. It would be great to add the requirement to the documentation as well.

fcojfernandez commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,226 @@
+ * 10. Check the {@link #stringInLogExpected()} is set during the restart.
+ *
+ * All the plugin author needs to do is override the methods providing:
+ * 1. The resource with the yaml configuration of the plugin
```
Now it's optional and it's not mandatory to override
```suggestion
 * 1. The resource with the yaml configuration of the plugin in case they use their own name for the file
```

casz commented on pom.xml:
```diff
@@ -93,6 +93,16 @@
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <excludes>
+            <exclude>**/ProjectMatrixAuthorizationTest.java</exclude>
```
WHY?

MRamonLeon replied:
Again, I pushed it by mistake, sorry. For some reason, these two tests are not running on my laptop. That's why I'm so obsessed with removing them.

MRamonLeon commented:
We have 3 plugins using this feature so far:
* https://github.com/jenkinsci/metrics-plugin/pull/46
* https://github.com/jenkinsci/ssh-credentials-plugin/pull/42
* https://github.com/jenkinsci/ssh-slaves-plugin/pull/126

MRamonLeon commented:
@casz is there anything else I can do to get it merged? Although this test has hacks it had helped a lot to find some bugs. Maybe it's good to merge it but avoid adding it to the documentation until the bugs are solved.

timja commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,310 @@
+            //String schema = getSchemaViaWebUI();
+
+            // Verify it's compliant to the schema
+            // TODO: when YAML schema validation is mature enough we can do that.
```
I think we can remove all the commented out code, make's it more difficult to see what's actually being done in the present.
YAML schema is a long way off working its current state

MRamonLeon replied:
ok

timja commented:
If you're happy with this please remove WIP

MRamonLeon commented:
The test using the abstract class and checking the result of the export against the YAML used is green:
![Screenshot from 2019-06-06 11-30-37](https://user-images.githubusercontent.com/22069922/59022748-ad6ded80-884e-11e9-8976-130a8124dd63.png)

Now, I'm going to comment the check out to avoid tests with secrets to fail.

MRamonLeon commented:
It would be great if we can merge and release it. Is there anything else I can do to achieve that?

oleg-nenashev commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,286 @@
+ * All the plugin author needs to do is override the methods providing:
+ * 1. The resource with the yaml configuration of the plugin in case they use their own name for the file.
+ * 2. A way to validate the configuration is established.
+ * 3. A string that should be present in the logs that guarantees the config is loaded. Usually a weird text configured.
```
```suggestion
 * 3. A string that should be present in the logs that guarantees the config is loaded. Usually a weird text configured.
 *
 * @since TODO
```

oleg-nenashev commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,286 @@
+        });
+    }
+
+//    private void assertJenkinsHasConfiguredPath(String jenkinsConfig, String resourcePath) throws IOException, URISyntaxException {
```
If you comment out a big chunk of code, please put explanation, or a link to the discussion in comments before it.
Bonus points for TODO

MRamonLeon replied:
done

oleg-nenashev commented on plugin/src/test/java/io/jenkins/plugins/casc/misc/ExportImportRoundTripAbstractTest.java:
```diff
@@ -0,0 +1,286 @@
+        assertThat(res, containsString("The configuration can be applied"));
+    }
+
+    private void applyConfigViaWebUI(String jenkinsConfig) throws Exception {
```
Maybe later we could have JCasC methods, e,g, in https://github.com/jenkinsci/java-client-api

MRamonLeon replied:
it would be great

MRamonLeon commented:
@timja, @varyvol I don't want to lose these pieces of code. We only need to uncomment that after fixing bugs. 

If merging without squashing, I'll be happy to remove all the commented out code.

timja commented:
Your PR will always be here and we can resurrect the code from this, we can do a rebase and merge though

casz commented:
Squash and merge does not squash the commits inside the PR, and your branch and PR will still have the code and history as @timja said.
So I'd prefer if you could remove the commented code 😓

casz commented:
@MRamonLeon are we good to merge? 👍

MRamonLeon commented:
I've removed the commented out code, although now the name of the class and method is not accurate because it's not doing an export/import. I've changed that as well because for now, I can change all the existing implementations.

MRamonLeon commented:
uuuooohhhh you were hidden under the curtains :laughing:

MRamonLeon commented:
Absolutely

casz commented:
We are too fast for you 😅 I suggest using [refined-github](https://github.com/sindresorhus/refined-github) as it will do the magic and pull new in comments.

casz commented:
Waiting on CI 😓

MRamonLeon commented:
aham, so there was a trick, it wasn't just...
![image](https://user-images.githubusercontent.com/22069922/59085486-0c8b3b00-88ff-11e9-8463-49f9a75ef45d.png)

timja commented:
![image](https://user-images.githubusercontent.com/21194782/59085747-8c150c00-88f7-11e9-9c77-417301ed575f.png)
@casz

casz commented:
I know @timja not so useful when sitting in Train on mobile data... 😓 
Kinda want to remove LGTM it is simply too slow...

MRamonLeon commented:
Ready!

MRamonLeon commented:
Can you cut off a release, to avoid incremental in the plugins extending the test? Thank you.

oleg-nenashev commented:
@casz if you are fine, I could use it as an opportunity to test the release flow on my machine

casz commented:
@oleg-nenashev I think we are ready, I would ask you setup GPG signing of tags at the very least but going through with a release. 😓

casz commented:
@MRamonLeon @fcojfernandez perhaps the missing things could be fixed now?
